elasticdl/doc/custom_data_source_support_design.md

```diff
@@ -0,0 +1,77 @@
+## Design for Custom Data Source Support
+
+This document describes the design for supporting custom data source in ElasticDL.
+
+### Support Custom Data Source in ElasticDL
+
+There are a couple of places in the the current codebase (as of commit [77cc87a](https://github.com/sql-machine-learning/elasticdl/tree/77cc87a90eec54db565849f0ae07d271fd957190))
+that are coupled heavily with RecordIO, for example:
+
+* When we create and dispatch tasks via ``_make_task_dispatcher()``, we rely on scanning directories that contain
+RecordIO files and reading the index for each file to obtain dictionaries of `{file_path: num_records}`.
+We then use this information to create tasks.
+* ``Worker`` starts a `TaskDataService` that generates a `tf.data.Dataset` instance that contains the actual data.
+The data is fetched in `TaskDataService._gen()` by opening a RecordIO reader to read the data offsets that each task
+we previously created.
+
+In order to support custom data sources in ElasticDL, we propose to make the following changes:
+
+* Rename ``shard_file_name`` in `Task` protobuf definition to `shard_name` so a task is independent with file names and `shard_name` will
+serve as an identifier for the shard.
+* Implement an abstract interface named `AbstractDataReader` that users can implement in their model definition file.
+The interface would look like the following:
+
+```python
+class AbstractDataReader(object):
+    def __init__(self, *kwargs):
+        pass
+
+    @abstractmethod
+    def read_records(self, task):
+        """This method will be used in `TaskDataService` to read the records based on
+        the information provided for a given task into a Python generator/iterator.
+        """
+        pass
+
+    @abstractmethod
+    def create_task_dispatcher(self):
```

Comment:
May need to change the func name. At the first glance, I thought users need to handle the creation of training/evaluation tasks. After seeing the following example, it looks like just a dict describing how many records inside each shard. In that case, why not just one AbstractDataReader maps to one dataset, and we don't mix training/evaluation/prediction together?

Reply:
Good idea! The creation of `TaskDispatcher` object that's returned in this function should not be done by the user. 

Three `AbstractDataReader`s will increase the number of CLI args that we need to add to the client. However about three separate methods in this class, e.g. `create_training_shards`, `create_evaluation_shards`, and `create_prediction_shards`?

Reply:
`create_shards` with a mode defined in `constants.Mode`?